go/collector.go

```diff
@@ -185,11 +212,72 @@ func (c *Collector) throttle() {
 	}
 }
 
-func (c *Collector) uploadToThunderstorm(info infoWithPath) (redo bool) {
+func (c *Collector) uploadToThunderstorm(info *infoWithPath) (redo bool) {
+	if c.shouldSkipFilePreOpen(info) {
+		return
+	}
+
+	f, err := os.Open(info.path)
+	if err != nil {
+		c.logger.Printf("Could not open file %s: %v\n", info.path, err)
+		atomic.AddInt64(&c.Statistics.fileErrors, 1)
+		return
+	}
+	defer f.Close()
+
+	if c.shouldSkipFilePostOpen(info, f) {
+		return
+	}
+
+	c.throttle()
+
+	contentType, formData := c.getFileContentAsFormData(f, info.path)
+	response, err := http.Post(c.thunderstormUrl(), contentType, formData)
+	if err != nil {
+		if info.retries < 3 {
+			c.logger.Printf("Could not send file %s to thunderstorm, will try again: %v", info.path, err)
+			info.retries++
+			time.Sleep(time.Second)
+			return true
+		} else {
+			c.logger.Printf("Could not send file %s to thunderstorm, canceling it.", info.path)
+			atomic.AddInt64(&c.Statistics.uploadErrors, 1)
+			return false
+		}
+	}
+	defer response.Body.Close()
+	if response.StatusCode == http.StatusServiceUnavailable {
+		retryAfter := response.Header.Get("Retry-After")
+		retryTime, err := strconv.Atoi(retryAfter)
+		if err != nil {
+			retryTime = 30 // Default to 30 seconds cooldown time
+		}
+		c.logger.Printf("Thunderstorm has no free capacities for file %s, retrying in %d seconds", info.path, retryTime)
+		time.Sleep(time.Second * time.Duration(retryTime))
+		return true
+	}
+	responseBody, _ := ioutil.ReadAll(response.Body)
+	if response.StatusCode != http.StatusOK {
+		atomic.AddInt64(&c.Statistics.uploadErrors, 1)
+		c.logger.Printf("Received error from Thunderstorm for file %s: %d %v\n", info.path, response.StatusCode, string(responseBody))
+		return
+	}
+	atomic.AddInt64(&c.Statistics.uploadedFiles, 1)
+	if c.Sync {
+		c.logger.Printf("Response to file %s: %s", info.path, string(responseBody))
+	}
+
+	if c.Debug {
+		c.debugf("File %s processed successfully", info.path)
+	}
+	return
+}
+
+func (c *Collector) shouldSkipFilePreOpen(info *infoWithPath) bool {
```

Comment:
shouldSkipFilePreOpen / shouldSkipFilePostOpen aren't really good names (they describe when the methods are called, but not what they do / where the difference is).

Maybe rename them to be more like "isUninterestingMetadata / isUninterestingContent" or something like that?

Reply:
Sounds good to me :+1: